php-checks/src/main/java/org/sonar/php/checks/GenericExceptionCheck.java

```diff
@@ -48,87 +48,90 @@
-    }
-    if (isBaseException(astNode)) {
-      getContext().createLineViolation(this, MESSAGE, astNode);
+  public void visitThrowStatement(ThrowStatementTree tree) {
+    NamespaceNameTree namespaceName = getNamespaceName(tree);
+
+    if (namespaceName != null && isGenericException(namespaceName)) {
+      context().newIssue(this.KEY, MESSAGE).tree(namespaceName);
     }
-  }
 
-  private void parseNamespace(AstNode astNode) {
-    AstNode namespaceNode = astNode.getFirstChild(PHPGrammar.NAMESPACE_NAME);
-    namespace = getStringValue(namespaceNode);
+    super.visitThrowStatement(tree);
   }
 
-  private void parseUse(AstNode astNode) {
-    AstNode namespaceNode = astNode.getFirstChild(PHPGrammar.NAMESPACE_NAME);
-    String namespaceValue = getStringValue(namespaceNode);
-    for (String exc : RAW_EXCEPTIONS) {
-      if (namespaceValue.equals(exc)) {
-        AstNode idNode = astNode.getFirstChild(PHPGrammar.IDENTIFIER);
-        String alias = idNode != null ? getStringValue(idNode) : namespaceValue;
-        uses.add(alias);
+  @Override
+  public void visitUseClause(UseClauseTree tree) {
+    if (tree.alias() != null) {
+      uses.add(tree.alias().text());
+
+    } else {
+      String name = tree.namespaceName().name().text();
+      if (tree.namespaceName().namespaces().isEmpty() && RAW_EXCEPTIONS.contains(name)) {
+        uses.add(name);
       }
     }
+
+    super.visitUseClause(tree);
   }
 
-  private boolean isBaseException(AstNode throwNode) {
-    AstNode classNameNode = throwNode.getFirstDescendant(PHPGrammar.CLASS_NAME);
-    if (classNameNode == null) {
-      return false;
-    }
-    String className = getStringValue(classNameNode);
-    if (className.startsWith(SEPARATOR)) {
-      return RAW_EXCEPTIONS.contains(className.substring(1));
+  @Override
+  public void visitNamespaceStatement(NamespaceStatementTree tree) {
+    if (tree.namespaceName() != null) {
+      currentNamespace = tree.namespaceName().fullName();
     }
-    if (namespace.isEmpty()) {
-      return RAW_EXCEPTIONS.contains(className);
+
+    super.visitNamespaceStatement(tree);
+
+    // Delimited namespace with curly braces
+    if (tree.openCurlyBrace() != null) {
+      currentNamespace = "";
     }
+  }
+
+  private boolean isGenericException(NamespaceNameTree namespaceName) {
+    String name = namespaceName.name().text();
+
+    return uses.contains(name)
+      // Should be from the global namespace, otherwise it's a user class named as the native one
+      || (isFromGlobalNamespace(namespaceName) && RAW_EXCEPTIONS.contains(name));
 
-    return uses.contains(className);
   }
 
-  private String getStringValue(AstNode classNameNode) {
-    if (classNameNode == null) {
-      return "";
-    }
+  private boolean isFromGlobalNamespace(NamespaceNameTree namespaceName) {
+    return namespaceName.namespaces().isEmpty()
+      && (currentNamespace.isEmpty() || !currentNamespace.isEmpty() && namespaceName.absoluteSeparator() != null);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS864)